### PR TITLE
fix(proxy): parse form-encoded video edit/extension bodies after auth

### LINKED ISSUE #36487: [Bug]: POST /v1/videos/edits and /v1/videos/extensions 500 with "RuntimeError: Stream consumed" for any form-encoded request
`POST /v1/videos/edits` is documented as a supported OpenAI-compatible route in [docs.litellm.ai/docs/videos](https://docs.litellm.ai/docs/videos), which also states that "LiteLLM follows the OpenAI Video Generation API specification". OpenAI's spec for this endpoint accepts **both** `application/json` and `multipart/form-data`. The JSON form works and matches the example in those docs; **every form-encoded request returns 500.**

## What happened

`video_edit` and `video_extension` read the raw request body with `await request.body()`, but `user_api_key_auth` has already consumed the ASGI stream by calling `await request.form()`. Starlette then raises `RuntimeError("Stream consumed")` and the proxy returns a 500 with no useful message.

This makes the endpoints unreachable from the official OpenAI Python SDK, because `client.videos.edit()` sets `Content-Type: multipart/form-data` unconditionally.

## Relevant code

`litellm/proxy/video_endpoints/endpoints.py`:

```python
# line 763, video_edit (identical shape at line 863 in video_extension, line 455 in video_remix)
body = await request.body()
data = orjson.loads(body)
```

The consuming read is `litellm/proxy/auth/user_api_key_auth.py:2529` → `_read_request_body()` → `litellm/proxy/common_utils/http_parsing_utils.py:64-66`:

```python
if _is_form_content_type(content_type):
    form_data = await request.form()   # drains the stream; _body is never cached
```

`video_generation` (line 82) and `video_create_character` (line 556) in the same file already use `await _read_request_body(request=request)` and are unaffected. `_read_request_body` is already imported at line 12.

## Minimal reproduction

`stub.py` — stands in for any OpenAI-compatible video backend:

```python
from fastapi import FastAPI
app = FastAPI()

@app.post("/v1/videos/edits")
async def edit():
    return {"id": "video_123", "object": "video", "created_at": 0,
            "status": "completed", "model": "sora-2", "progress": 100,
            "seconds": "4", "size": "720x1280"}
```

`config.yaml`:

```yaml
model_list:
  - model_name: my-video-model
    litellm_params:
      model: openai/my-video-model
      api_base: http://127.0.0.1:8111/v1
      api_key: none
general_settings:
  master_key: sk-1234
```

```bash
uvicorn stub:app --port 8111 &
litellm --config config.yaml --port 4000 &
```

```bash
# JSON -> 200 OK
curl -sS -X POST http://127.0.0.1:4000/v1/videos/edits \
  -H "Authorization: Bearer sk-1234" -H "Content-Type: application/json" \
  -d '{"model":"my-video-model","prompt":"brighter","video":{"id":"video_123"}}'

# form-encoded -> 500. Simplest trigger; no multipart or file needed.
curl -sS -X POST http://127.0.0.1:4000/v1/videos/edits \
  -H "Authorization: Bearer sk-1234" \
  -H "Content-Type: application/x-www-form-urlencoded" \
  --data-urlencode model=my-video-model \
  --data-urlencode prompt=brighter \
  --data-urlencode video=video_123
```

Or via the official SDK, which always sends multipart on this route:

```python
client.videos.edit(prompt="brighter",
                   video=("s.mp4", open("s.mp4","rb").read(), "video/mp4"),
                   extra_body={"model": "my-video-model"})
# openai.InternalServerError: Error code: 500
```

## Expected

The request is parsed and forwarded, as it is for `POST /v1/videos` and `POST /v1/videos/characters`.

## Actual

`500 {"error":{"message":"Internal server error","type":"internal_server_error"}}`, with this server-side traceback:

```
proxy_server.py:1484 - Unhandled exception in request: RuntimeError
  File "litellm/proxy/video_endpoints/endpoints.py", line 763, in video_edit
    body = await request.body()
  File "starlette/requests.py", line 257, in body
    async for chunk in self.stream():
  File "starlette/requests.py", line 240, in stream
    raise RuntimeError("Stream consumed")
RuntimeError: Stream consumed
```

## Versions

litellm 1.95.0 and 1.96.0, and `main` as of today (lines 455 / 763 / 863). starlette 1.3.1, fastapi 0.136.3, python 3.13. Also reproduced on starlette 1.3.1 + fastapi 0.136.3 + python 3.12.

## Suggested patch

`video_edit` (and the same shape at line 863 for `video_extension`):

```diff
- body = await request.body()
- data = orjson.loads(body)
-
- # Extract video_id from nested video object
- video_ref = data.pop("video", {})
- video_id = video_ref.get("id", "") if isinstance(video_ref, dict) else ""
+ data = await _read_request_body(request=request)
+
+ # Extract video_id from nested video object. Over multipart/form-data the
+ # reference arrives as a string: either JSON-encoded or a bare id.
+ video_ref = data.pop("video", {})
+ if isinstance(video_ref, dict):
+     video_id = video_ref.get("id", "")
+ elif isinstance(video_ref, str):
+     try:
+         parsed_ref = orjson.loads(video_ref)
+     except orjson.JSONDecodeError:
+         parsed_ref = None
+     video_id = parsed_ref.get("id", "") if isinstance(parsed_ref, dict) else video_ref
+ else:
+     video_id = ""
  data["video_id"] = video_id
```

Verified locally: all four shapes (JSON reference, multipart form field, multipart JSON-encoded reference, and a bare id string) return 200 and normalize to the identical outbound body `{"prompt":"brighter","video":{"id":"video_123"}}`.

## Two follow-ups worth splitting out

1. Without the `str` branch, a non-dict `video` is silently coerced to `{"id": ""}` and forwarded. A malformed reference should be a 400, not a downstream failure with a blank id.

2. OpenAI's spec allows `video` in `CreateVideoEditMultipartBody` to be `type: string, format: binary` — an uploaded source video. LiteLLM cannot express this: `litellm.videos.main.video_edit(video_id: str, prompt: str, ...)` takes no file, and `OpenAIVideoConfig.transform_video_edit_request` hardcodes `{"video": {"id": ...}}`. With the patch above, an SDK file upload returns 200 while the file is silently discarded. Until file uploads are plumbed through, this case should return an explicit 400 rather than succeed with an empty id.

## Related

[#27836](https://github.com/BerriAI/litellm/issues/27836) appears to be the same root cause in the A2A route. [#23737](https://github.com/BerriAI/litellm/pull/23737) is the PR that added these endpoints.

### PR DESCRIPTION
**Pull request by Souravrajvi0**
## TLDR

Problem this solves:

- Form-encoded and multipart video edit/extension requests return 500 on the proxy
- OpenAI SDK `client.videos.edit()` fails against LiteLLM with internal server error
- JSON-only video edit requests worked, so the bug looked like an SDK or provider issue

How it solves it:

- Read the request through the same cached body path auth already uses
- Normalize `video` from dict, bare id string, or JSON-encoded string into `video_id`
- Apply the same body handling to `video_remix` for consistency

## User Flow

Before: a developer editing a video through the OpenAI SDK gets a 500 from the LiteLLM proxy on every attempt

1. Their app calls `client.videos.edit(prompt="Make it brighter", video=..., extra_body={"model": "my-video-model"})` against `https://litellm-domain/v1`
2. The SDK sends `POST https://litellm-domain/v1/videos/edits` with `Content-Type: multipart/form-data`
3. The proxy responds `500` with `{"error":{"message":"Internal server error","type":"internal_server_error"}}`
4. The video edit job never starts and no usable video id is returned

After: the same SDK call is accepted and returns a normal video job response

1. Their app calls the same `client.videos.edit(...)` against `https://litellm-domain/v1`
2. The SDK sends the same `POST https://litellm-domain/v1/videos/edits` with multipart form data
3. The proxy responds `200` with a video object (`id`, `status`, `object: "video"`, etc.)
4. Their app can poll or display the returned video id like a direct OpenAI call

## Relevant issues

Fixes #36487

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have added meaningful tests
- [x] The handful of test files covering my change pass locally, e.g. `uv run pytest tests/test_litellm/<your_test_file>.py -v`. Leave the suites (`make test-unit-*`, `make test-unit`) to CI: it finishes in ~15 minutes where a laptop takes an hour or more
- [x] My PR passes all required CI/CD checks (e.g., lint, schema.d.ts sync check, etc.)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [ ] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review (Greptile reviews automatically once the PR is opened; only comment `@greptileai` to re-request a review after pushing changes)

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Setup, identical on both sides: an OpenAI-compatible video backend stub on port 8931, and a proxy with this `config.yaml` (Before on port 8932, After on port 8933):

```yaml
model_list:
  - model_name: my-video-model
    litellm_params:
      model: openai/my-video-model
      api_base: http://127.0.0.1:8931/v1
      api_key: none
general_settings:
  master_key: sk-1234
```

#### JSON edits (sanity, always worked)

1. `curl -sS -X POST http://127.0.0.1:8932/v1/videos/edits -H "Authorization: Bearer sk-1234" -H "Content-Type: application/json" -d '{"model":"my-video-model","prompt":"brighter","video":{"id":"video_123"}}'`
2. `HTTP 200` with `{"id":"video_bGl0ZWxsbTpjdXN0b21fbGxtX3Byb3ZpZGVyOm9wZW5haTs...","object":"video","status":"completed",...}`

#### form-urlencoded edits

1. `curl -sS -X POST http://127.0.0.1:8932/v1/videos/edits -H "Authorization: Bearer sk-1234" -H "Content-Type: application/x-www-form-urlencoded" --data-urlencode model=my-video-model --data-urlencode prompt=brighter --data-urlencode video=video_123`
2. `HTTP 500` with `{"error":{"message":"Internal server error","type":"internal_server_error"}}`; proxy log shows `RuntimeError: Stream consumed`

#### multipart edits

1. `curl -sS -X POST http://127.0.0.1:8932/v1/videos/edits -H "Authorization: Bearer sk-1234" -F model=my-video-model -F prompt=brighter -F video=video_123`
2. `HTTP 500` with `{"error":{"message":"Internal server error","type":"internal_server_error"}}`

#### form-urlencoded extensions

1. `curl -sS -X POST http://127.0.0.1:8932/v1/videos/extensions -H "Authorization: Bearer sk-1234" -H "Content-Type: application/x-www-form-urlencoded" --data-urlencode model=my-video-model --data-urlencode prompt=continue --data-urlencode seconds=4 --data-urlencode video=video_123`
2. `HTTP 500` with `{"error":{"message":"Internal server error","type":"internal_server_error"}}`

#### OpenAI SDK multipart edit (openai 3.3.1)

1. `client = OpenAI(api_key="sk-1234", base_url="http://127.0.0.1:8932/v1")` then `client.videos.edit(prompt="Make it brighter", video=Path("fake.mp4"), extra_body={"model": "my-video-model"})`
2. `openai.InternalServerError: Error code: 500 - {'error': {'message': 'Internal server error', 'type': 'internal_server_error'}}`

#### JSON edits (sanity, always worked)

1. Same curl against port 8933
2. `HTTP 200` with the same video object

#### form-urlencoded edits

1. Same curl against port 8933
2. `HTTP 200` with `{"id":"video_bGl0ZWxsbTpjdXN0b21fbGxtX3Byb3ZpZGVyOm9wZW5haTs...","object":"video","status":"completed",...}` and `video_id` normalized to `video_123` for routing

#### multipart edits

1. Same curl against port 8933
2. `HTTP 200` with the same video object

#### form-urlencoded extensions

1. Same curl against port 8933
2. `HTTP 200` with the video object

#### OpenAI SDK multipart edit (openai 3.3.1)

1. Same SDK call against port 8933
2. Returns a `Video` object: `id="video_bGl0ZWxsbTpjdXN0b21fbGxtX3Byb3ZpZG..."`, `status="completed"`

## Type

🐛 Bug Fix

## Caveats (if any)

- Uploading a source video file as the `video` field is still not forwarded upstream (pre-existing gap noted in #36487)
- A malformed non-dict `video` reference may still become an empty id instead of a 400 (follow-up in #36487)
- Requests missing `seconds` on `/v1/videos/extensions` still 500 instead of 400 (pre-existing, JSON and form alike)

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

- [x] 603d3358d6b12216c834f236fda1c9c3de3aeda6 passes /live-pr-risk

---

> [!NOTE]
> **Low Risk**
> Scoped to proxy video edit/extension/remix request parsing; uses the same cached-body helper as other endpoints, with regression tests for form bodies and stream reuse.
> 
> **Overview**
> Fixes **500 / "Stream consumed"** failures on video **edit**, **extension**, and **remix** when clients send **form-urlencoded or multipart** bodies (e.g. OpenAI SDK `videos.edit()`), by replacing raw `request.body()` + JSON parsing with the shared **`_read_request_body`** path that reuses the body cached during API-key auth.
> 
> For **edit** and **extension**, **`video`** is normalized via new **`video_reference_to_id`**: nested `{"id": ...}` (JSON), a bare id string, or a JSON-encoded reference string become **`video_id`** before provider/model routing. **`video_remix`** uses the same body reader for consistency.
> 
> Tests cover form-style `video` values, **`video_reference_to_id`** edge cases, and handlers reading the cached body after an auth-style pre-read (no stream double-consume).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 603d3358d6b12216c834f236fda1c9c3de3aeda6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by mateo-berri** [comment]
@greptileai

**Comment by mateo-berri** [comment]
bugbot run

File: litellm/proxy/video_endpoints/utils.py [hunk 1]
```diff
@@ -13,6 +13,28 @@ def extract_model_from_target_model_names(target_model_names: Any) -> str | None
...
+    try:
+        parsed_ref: Final = orjson.loads(video_ref)
+    except orjson.JSONDecodeError:
+        return video_ref
+    return parsed_ref.get("id", "") if isinstance(parsed_ref, dict) else video_ref
+
+
+def pop_video_reference_to_video_id(data: dict[str, Any]) -> None:
+    """
+    Normalize OpenAI video edit/extension payloads into ``video_id``.
```
[the remark this answers is not in the document]
Comment (mateo-berri) [reply litellm/proxy/video_endpoints/utils.py]: Fixed both: video_reference_to_id is now a pure function, and the regression test calls the video_edit/video_extension handlers directly instead of patching a class method

**Comment by mateo-berri** [comment]
@greptileai

**Comment by mateo-berri** [comment]
bugbot run

**Comment by mateo-berri** [comment]
@greptileai

**Review by mateo-berri — APPROVED** [review]
LGTM. Thanks for the contribution!
